lib/puppet/util/windows/process.rb

```diff
@@ -97,6 +121,32 @@ def with_process_token(access, &block)
   end
   module_function :with_process_token
 
+  def get_process_image_name_by_pid(pid)
+    image_name = ""
+
+     open_process(PROCESS_QUERY_INFORMATION, false, pid) do |phandle|
+
+       FFI::MemoryPointer.new(:dword, 1) do |exe_name_length_ptr|
+        # Add 1 for the null terminator, and UTF is 2 bytes/char:
+        max_path_length = (MAX_PATH_LENGTH + 1) * 2
+        exe_name_length_ptr.write_dword(max_path_length)
```

Comment:
According to [msdn docs](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-queryfullprocessimagenamew), `lpdwSize` needs to be characters not bytes: "On input, specifies the size of the lpExeName buffer, in characters". It's not clear to me whether it's supposed to include the null terminator or not. Probably safer to do `max_path_length = MAX_PATH_LENGTH`.

Reply:
This doesn't look quite right as `lpdwSize` is supposed to be the size of the buffer in wide characters, not bytes. As it is currently, we're claiming the buffer can accept `MAX_PATH_LENGTH * 2` characters, but only constructing a buffer of length `MAX_PATH_LENGTH * 2` bytes.

I think you'll want to do something like:

```ruby
open_process(PROCESS_QUERY_INFORMATION, false, pid) do |phandle|
  FFI::MemoryPointer.new(:dword, 1) do |exe_name_length_ptr|
    max_chars = MAX_PATH_LENGTH + 1
    exe_name_length_ptr.write_dword(max_chars)
    FFI::MemoryPointer.new(:wchar, max_chars) do |exe_name_ptr|
      use_win32_path_format = 0
      result = QueryFullProcessImageNameW(phandle, use_win32_path_format, exe_name_ptr, exe_name_length_ptr)
      ...
      image_name = exe_name_ptr.read_wide_string(exe_name_length_ptr.read_dword)
    end
  end
end
```

Note the buffer needs to be one wchar more than MAX_PATH_LENGTH, otherwise, if the process path is >= `MAX_PATH_LENGTH`, then `QueryFullProcessImageNameW` will fail with 0 and `GetLastError` will return `ERROR_INSUFFICIENT_BUFFER`.